distributed/shuffle/multi_file.py

```diff
@@ -0,0 +1,282 @@
+from __future__ import annotations
+
+import asyncio
+import contextlib
+import os
+import pathlib
+import pickle
+import shutil
+import time
+import weakref
+from collections import defaultdict
+
+from dask.sizeof import sizeof
+from dask.utils import parse_bytes
+
+from distributed.utils import log_errors
+
+
+class MultiFile:
+    """Accept, buffer, and write many small objects to many files
+
+    This takes in lots of small objects, writes them to a local directory, and
+    then reads them back when all writes are complete.  It buffers these
+    objects in memory so that it can optimize disk access for larger writes.
+
+    **State**
+
+    -   shards: dict[str, list[bytes]]
+
+        This is our in-memory buffer of data waiting to be sent to other workers.
+
+    -   sizes: dict[str, int]
+
+        The size of each list of shards.  We find the largest and send data from that buffer
+
+    State
+    -----
+
+    memory_limit: str
+        A maximum amount of memory to use, like "1 GiB"
+    max_connections: int
+        The maximum number of connections to have out at once
+    max_message_size: str
+        The maximum size of a single message that we want to send
+
+    Parameters
+    ----------
+    directory: pathlib.Path
+        Where to write and read data.  Ideally points to fast disk.
+    dump: callable
+        Writes an object to a file, like pickle.dump
+    load: callable
+        Reads an object from that file, like pickle.load
+    send: callable
+        How to send a list of shards to a worker
+    sizeof: callable
+        Measures the size of an object in memory
+    """
+
+    memory_limit = parse_bytes("1 GiB")
+    _queues: weakref.WeakKeyDictionary = weakref.WeakKeyDictionary()
+    concurrent_files = 2
+    total_size = 0
+
+    def __init__(
+        self,
+        directory,
+        dump=pickle.dump,
+        load=pickle.load,
+        sizeof=sizeof,
+        loop=None,
+    ):
+        self.directory = pathlib.Path(directory)
+        if not os.path.exists(self.directory):
+            os.mkdir(self.directory)
+        self.dump = dump
+        self.load = load
+        self.sizeof = sizeof
+
+        self.shards = defaultdict(list)
+        self.sizes = defaultdict(int)
+        self.total_size = 0
+        self.total_received = 0
+
+        self.condition = asyncio.Condition()
+
+        self.bytes_written = 0
+        self.bytes_read = 0
+
+        self._done = False
+        self._futures = set()
+        self.active = set()
+        self.diagnostics = defaultdict(float)
+
+        self._communicate_future = asyncio.ensure_future(self.communicate())
+        self._loop = loop or self
```

Comment:
This looks... wrong?

Reply:
I added this when I started adding testing.  We need some object here that serves as a unique identifier for event loops.  I didn't want to have to create event loop infrastructure just for the tests.  I can though.   Hopefully asyncio cleans them up properly in between tests.